### PR TITLE
Improvement: Production Config and Makefile improvements

### PR DESCRIPTION
**Pull request by snova-kwasia**

### CODE REVIEW

File: utils/prod/update_config.py [hunk 1]
```diff
@@ -29,48 +30,69 @@
...
+    for key in keys:
+        if isinstance(data, dict) and key in data:
+            data = data[key]
+        else:
+            return None
+    return data if isinstance(data, dict) else None
+
+
+def update_nested_dict(data: Dict[str, Any], keys: List[str], value: Any) -> None:
+    """Update a nested dictionary given a list of keys and a value."""
```
Comment (snova-francescar) [inline utils/prod/update_config.py:44]: One line full space after the docstring.

File: utils/prod/update_config.py [hunk 2]
```diff
@@ -29,48 +30,69 @@
...
+    for key in keys[:-1]:
+        data = data.setdefault(key, {})
+    data[keys[-1]] = value
```
Comment (snova-francescar) [inline utils/prod/update_config.py:47]: Ok so this is what happens to the last key. Why?
We need comments here to understand.

File: utils/prod/update_config.py [hunk 3]
```diff
@@ -29,48 +30,69 @@
...
+
+
 def update_embedding_model(config: Dict[str, Any]) -> None:
     """Update the embedding model configuration."""
```
Comment (snova-francescar) [inline utils/prod/update_config.py:51]: Space below, because it is a one liner.

File: utils/prod/update_config.py [hunk 4]
```diff
@@ -29,48 +30,69 @@
...
-    if 'embedding_model' in config:
-        embedding_model = config['embedding_model']
-        if isinstance(embedding_model, dict):
-            embedding_model['type'] = 'sambastudio'
-            embedding_model['batch_size'] = 32
-            embedding_model['coe'] = False
+    embedding_model = get_nested_dict(config, ['rag', 'embedding_model']) or get_nested_dict(
```
Comment (snova-francescar) [inline utils/prod/update_config.py:52]: More comments here to explain what we are retrieving.

File: utils/prod/update_config.py [hunk 5]
```diff
@@ -29,48 +30,69 @@
...
+        config, ['embedding_model']
+    )
+    if embedding_model:
```
Comment (snova-francescar) [inline utils/prod/update_config.py:55]: So if we are using the embedding models, we overwrite the config parameters with the sambastudio ones (similarly to what we do on `replit`).
We need to explain that.

File: utils/prod/update_config.py [hunk 6]
```diff
@@ -29,48 +30,69 @@
...
+        embedding_model['type'] = 'sambastudio'
+        embedding_model['batch_size'] = 32
+        embedding_model['coe'] = False
 
 
 def update_config_prod(config: Dict[str, Any]) -> None:
     """Update configuration for production mode."""
```
Comment (snova-francescar) [inline utils/prod/update_config.py:62]: Space.

File: utils/prod/update_config.py [hunk 7]
```diff
@@ -29,48 +30,69 @@
...
-    if 'prod_mode' in config:
-        config['prod_mode'] = True
-
+    update_nested_dict(config, ['prod_mode'], True)
```
Comment (snova-francescar) [inline utils/prod/update_config.py:63]: Here we enforce the prod mode. We should explain why.

File: utils/prod/update_config.py [hunk 8]
```diff
@@ -29,48 +30,69 @@
...
     update_embedding_model(config)
 
-    if 'audio_model' in config:
-        audio_model = config['audio_model']
-        if isinstance(audio_model, dict):
-            audio_model['model'] = 'whisper-1'
+    audio_model = get_nested_dict(config, ['audio_model'])
+    if audio_model:
+        audio_model['model'] = 'whisper-1'
```
Comment (snova-francescar) [inline utils/prod/update_config.py:68]: Again, overwrite here needs explanation.

File: utils/prod/update_config.py [hunk 9]
```diff
@@ -29,48 +30,69 @@
...
+
+    # Update pages_to_show for prod mode
+    update_nested_dict(config, ['pages_to_show'], ['synthetic_eval'])
 
 
-def update_makefile(root_dir: str) -> None:
+def update_makefile(root_dir: str, port: int) -> None:
     """Update the STREAMLIT_PORT in the Makefile."""
```
Comment (snova-francescar) [inline utils/prod/update_config.py:75]: Space after.

File: utils/prod/update_config.py [hunk 10]
```diff
@@ -29,48 +30,69 @@
...
     makefile_path = os.path.join(root_dir, 'Makefile')
     if os.path.exists(makefile_path):
-        with open(makefile_path, 'r') as file:
+        with open(makefile_path, 'r', encoding='utf-8') as file:
             makefile_content = file.read()
 
         new_makefile_content = re.sub(
-            r'^(STREAMLIT_PORT\s*:=\s*)\d+', r'\g<1>443', makefile_content, flags=re.MULTILINE
+            r'^(STREAMLIT_PORT\s*:=\s*)\d+',
```
Comment (snova-francescar) [inline utils/prod/update_config.py:82]: Comments.

File: utils/prod/update_config.py [hunk 11]
```diff
@@ -29,48 +30,69 @@
...
             file.write(new_makefile_content)
-        print(f'Updated STREAMLIT_PORT to 443 in {makefile_path}')
+        print(f'Updated STREAMLIT_PORT to {port} in {makefile_path}')
     else:
         print(f'Makefile not found at {makefile_path}. Skipping STREAMLIT_PORT update.')
 
 
-def update_configs(mode: str = 'prod') -> None:
+def update_configs(mode: str = 'prod', port: int = 8501) -> None:
     """Update configurations for all kits based on the specified mode."""
```
Comment (snova-francescar) [inline utils/prod/update_config.py:96]: Space after.

File: utils/prod/update_config.py [hunk 12]
```diff
@@ -83,33 +105,39 @@ def update_configs(mode: str = 'prod') -> None:
...
 
         if mode == 'prod':
             update_config_prod(config)
         else:
             update_embedding_model(config)
 
-        with open(config_path, 'w') as file:
-            yaml.dump(config, file)
+        with open(config_path, 'w', encoding='utf-8') as file:
+            yaml.dump(config, file, default_flow_style=False)
```
Comment (snova-francescar) [inline utils/prod/update_config.py:117]: We need comments to explain why we don't want the default style.

File: utils/prod/update_config.py [hunk 13]
```diff
@@ -83,33 +105,39 @@ def update_configs(mode: str = 'prod') -> None:
...
-if __name__ == '__main__':
-    import sys
+def main() -> None:
+    """Main function to parse arguments and run the update process."""
+    parser = argparse.ArgumentParser(description='Update configurations for AI Starter Kits.')
+    parser.add_argument(
+        'mode',
+        nargs='?',
+        default='prod',
+        choices=['prod', 'test'],
```
Comment (snova-francescar) [inline utils/prod/update_config.py:132]: Why choosing the name `test` here?
Is test mode the standard (as opposed to prod mode)? Why not dev mode?

File: tests/test_framework.py [hunk 14]
```diff
@@ -91,6 +98,7 @@ class StarterKitTest(unittest.TestCase):
     run_streamlit: ClassVar[bool]
     run_cli: ClassVar[bool]
     is_docker: ClassVar[bool]
+    recreate_venv: ClassVar[bool]
```
[the remark this answers is not in the document]
Comment (snova-kwasia) [reply tests/test_framework.py:101]: We're using ClassVar to indicate that these variables are class-level attributes shared across all instances of the class. This helps with type hinting and makes it clear that these attributes are not instance-specific.

File: Makefile [hunk 15]
```diff
@@ -51,14 +50,20 @@ TEST_SUITE_VENV := .test_suite_venv
 TEST_SUITE_REQUIREMENTS := tests/requirements.txt
 BASE_REQUIREMENTS := base-requirements.txt
 
+# Set default value for PARSING
```
[the remark this answers is not in the document]
Comment (snova-kwasia) [reply Makefile:53]: I agree that adding more comments would improve clarity. I'll create a follow-up PR dedicated to adding detailed comments throughout the Makefile

File: utils/prod/update_config.py [hunk 16]
```diff
@@ -29,48 +30,69 @@
...
 
 
+def get_nested_dict(data: Dict[str, Any], keys: List[str]) -> Optional[Dict[str, Any]]:
+    """Retrieve a nested dictionary given a list of keys."""
+    for key in keys:
+        if isinstance(data, dict) and key in data:
+            data = data[key]
+        else:
+            return None
+    return data if isinstance(data, dict) else None
```
[the remark this answers is not in the document]
Comment (snova-kwasia) [reply utils/prod/update_config.py:40]: The get_nested_dict function is designed to retrieve the nested dictionary corresponding to the last key in the list by traversing through the nested keys. It doesn't modify the original data but navigates through it. But have changed to copy for safer execution.

File: utils/prod/update_config.py (hunk 2 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:47]: The update_nested_dict function iterates through all keys except the last one to navigate to the appropriate nested dictionary, creating intermediate dictionaries if they don't exist (setdefault). Then, it updates the value of the last key.  Comments added

File: utils/prod/update_config.py (hunk 5 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:55]: Yes. We're standardizing the embedding model settings to use sambastudio, similar to our setup on hosted kits. I'll add comments to explain this overwrite for better clarity.

File: utils/prod/update_config.py (hunk 7 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:63]: Enforcing prod_mode ensures that the application runs with production settings, which might include optimizations and configurations different from development mode. I'll add a comment to explain the importance of this enforcement.

File: utils/prod/update_config.py (hunk 8 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:68]: Added

File: utils/prod/update_config.py (hunk 10 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:82]: Added

File: utils/prod/update_config.py (hunk 12 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:117]: Setting default_flow_style=False ensures that the YAML is written in block style, which is more readable. I'll add a comment to explain this choice.

File: utils/prod/update_config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (snova-francescar) [reply utils/prod/update_config.py:40]: But why are we not taking the last key directly? Why traversing the main dict?

File: utils/prod/update_config.py (hunk 13 above)
Comment (snova-kwasia) [reply utils/prod/update_config.py:132]: We used 'test' as this mode is for running our global test suite. Comment added

File: utils/prod/update_config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (snova-francescar) [reply utils/prod/update_config.py:40]: Then if one of the previous keys does not correspond to a dict, it returns `None`!

**Pull request merged**